### DISCUSSION TITLE
Idea: LNURL

### DISCUSSION THREAD

**Discussion by matthiasdebernardini** [fedimint/fedimint#1516] [Fedimint Modules]
LN-URL for federation guardians and for users.

Also add Lightning Addresses so that pay-to-email works.

**Comment by elsirion** [comment]
This will be hard/impossible to implement securely since to my knowledge LNURL responses aren't signed using a scheme that lends itself to be run as a threshold scheme.

For LNURL Pay recurring invoices to be secure the initial LNURL would need to commit to a pub key with which the generated invoices need to be threshold-signed (and this is widely enforced by clients). Currently LNURL's security model depends on TLS/PKI and assumes that users are talking to a single server that is trusted (which doesn't exist in Fedimint's security model).

Unless LNURL has been or will be patched this way BOLT12 seems the better alternative #1507.

**Reply by matthiasdebernardini** [reply]
For the record, not supporting LNURL means not supporting [Lightning Addresses](https://lightningaddress.com).

Sounds like either adapting BOLT12 to lightning addresses or patching LNURL are the way to go. 

I think not having support for Lightning Addresses is hugely detrimental for the UX/UI of fedimint users. So if its doable then it should be added.

**Reply by moneyball** [reply]
Bolt12 will enable an improved version of LNURL that doesn't require a web server request per payment. Instead, a one-time web server request can be made to receive a bolt12 offer (a static, reusable payment identifier). Client software can then easily map that bolt12 offer to whatever UI makes sense for the application (a nym user name, a real name, a phone #, a nostr pubkey, etc.). This address the privacy and censorship concerns around LNURL.

**Reply by ecurrencyhodler** [reply]
Question. I'm not sure where the latest documentation is on how fedimints will run lightning nodes. Apologies. But based off of eric's comment, it sounds like each lightning payment will need to be signed by multiple members in the federation?

**Reply by justinmoon** [reply]
> But based off of eric's comment, it sounds like each lightning payment will need to be signed by multiple members in the federation?

With BOLT12, the federation would be able to turn BOLT12 offers into BOLT12 invoices while end-user is offline. This is possible because the offer contains a schnorr pubkey, and the federation can produce signatures for this pubkey. This isn't possible with LNURL because it contains no such pubkey.

**Reply by moneyball** [reply]
More progress on BOLT 12 / DNS / human-readable names / Lightning Address compatibility:

https://github.com/bitcoin/bips/pull/1551
https://github.com/lightning/bolts/pull/1136
https://github.com/lightning/blips/pull/32
https://http-dns-prover.as397444.net](https://t.co/CvFleP1Vx8)
https://git.bitcoin.ninja/index.cgi?p=dnssec-prover;a=tree;h=refs/heads/master;hb=refs/heads/master

**Comment by m1sterc001guy** [comment]
I think a trusted version of LNURL-pay might be worth implementing in the gateway (would be easier than implementing BOLT12 currently). Gateway's already run an always on web server and invoices in our LNv2 are signed by the Gateway's Lightning node. It would be a really nice user experience for Fedimint users to automatically get an LN address just by using Fedimint.

Protocol might look something like this:

1. User buys/registers a username with a gateway (perhaps behind L402).
2. User uploads encrypted payment preimages, payment hashes, and user's claim public key to the gateway.
3. When the gateway needs to create an invoice, the gateway is responsible for creating the `IncomingContract` using one of the encrypted preimages/hashes and it keyed to the user's claim key (i.e only the user can claim the ecash). This is where the trusted part comes in, the user is trusting the gateway to create the `IncomingContract` with the correct amount. The user client cannot directly upload commitments, since the commitments in LNv2 also commit to the amount, which only the gateway knows at this time. Invoice is then returned to the sender.
4. The user client can then either work with the gateway to find the `IncomingContract` that it can sweep, or just scan all outstanding `IncomingContracts`. The user client would then claim the incoming ecash.

We have already been talking about adding some trusted functionality to the gateway (such as "melting" ecash) and I think this would be relatively straightforward to implement and a nice user experience.

Thoughts?

**Reply by TheBlueMatt** [reply]
What's the blocker for just shipping BOLT 12 and being done with it?

**Reply by m1sterc001guy** [reply]
@joschisan @elsirion, tnull (not sure why I can't tag), and I had a good discussion about this at btc++.

> What's the blocker for just shipping BOLT 12 and being done with it?

For Bolt 12, there are no issues with send support and that is on the roadmap. Receive support is trickier to implement in a trust minimized way, it essentially has the same problem that LNURL has that I described above. I am working on a summary of our btc++ conversation and I will post it to the BOLT12 discussion thread we have.

Outside of BOLT12 though, here are the reasons I think we should also do LNURL:

1. Unlike "regular" lightning, Fedimint already requires an always online web server (i.e the gateway) to service lightning payments. LNURL requires a webserver and since the gateway already has one, it seems like a good fit.
2. It would be easy to implement for all of our backends (LND, CLN, and LDK). We wouldn't need a separate implementation for each.
3. Most users use LNURL with a custodial wallet anyway. LNURL in Fedimint is an improvement (even with the above described trust assumption). In Fedimint, the custody is federated. With Fedimint LNURL, the gateway is not custodying your money (that is still done by the federation), you are only trusting the gateway to not steal your money (which I believe is the same trust assumption that any lightning user using LNURL has when they're not running their own webserver).
4. Ecash is being used more and more on Nostr and I would like to see Fedimint has participate in zaps. This is mostly a selfish reason :)

I want to be clear, I'm not advocating for LNURL over Bolt12, I think we should do both. Given that BOLT12 is, in my opinion, the superior protocol, I wouldn't really be comfortable implementing a trusted form of BOLT12 receive. I think this would really mislead users. I think we _need_ to implement a trustless BOLT12 receive, it's just going to take a bit more time. LNURL might be a decent stop-gap feature.

**Reply by TheBlueMatt** [reply]
Don't forget to implement sanctions compliance checks on the LNURL gateway! Need to bake in blocking Russia/Iran/North Korea/New York or people might go to prison. :)

Seriously, though, I don't fully understand why not implement a trusted BOLT 12 option. If your take is that users are going to be confused by the UX flow of using a QR code and expecting that to be self-custodial, that's fair, but if you're going to implement LN-Address for zaps today you probably need to also implement BIP 353 to have broad compatibility going forward, rather than starting out the gate with incompatibilities in your Human Readable Address that you expose to users. You could implement BOLT 12 trusted receive and only use it for BIP 353 rather than exposing the QR code for the offer (which seems premature at this juncture anyway given most senders don't yet support BOLT 12).

**Reply by elsirion** [reply]
Even the trusted BOLT12 impl requires some more sophistication (handling onion messages etc.) that we need to implement in a way compatible with our 3 LN backends and our custom fake channel protocol (at least for LNv1, which is currently deployed, LNv2 is still ~3M out, idk how this would change things). LNURL in comparison is just running a web server (and is already implemented outside the main Fedimint repo in the form of [Hermes](https://github.com/Kodylow/hermes)).

I think trusted BOLT12 would be useful just to get a feeling for the protocol and discovering misunderstandings that would lead to far more wasted work in the trust-minimized implementation.

**Reply by TheBlueMatt** [reply]
> Even the trusted BOLT12 impl requires some more sophistication (handling onion messages etc.) that we need to implement in a way compatible with our 3 LN backends and our custom fake channel protocol

Presumably you'd simply not implement it for the LND backend but only CLN/LDK ones. That should be very straightforward (if its trusted you don't need to bother with the fake channel protocol stuff, which only exists to make things non-trusted), probably even simpler than running a web server cause you can just add a new invoice format :).

**Reply by elsirion** [reply]
> if its trusted you don't need to bother with the fake channel protocol stuff

With LNURL we'd be generating invoices using the same fake channel protocol so that the user receives the money directly and the LNURL server never holds any money, they just assist in generating the invoice on behalf of the user. I think that's a much better place to be in as a service provider: you never assume custody of the funds.

Unless the BOLT12 version was to use a single custodian we'd have to run the same protocol.

**Reply by TheBlueMatt** [reply]
Right, you'd still have to build the logic to insert a custom payment hash in the BOLT 12 invoice, but you wouldn't need to have any BOLT 12 specific logic on the federation (unlike the trustless version).

**Reply by elsirion** [reply]
Yes, we'd need to find a way to intercept BOLT12 invoice requests. We created a tracking issue for the things we can do for BOLT12 support, mostly in ascending difficulty: https://github.com/fedimint/fedimint/issues/6223

**Reply by moneyball** [reply]
Will this PR unblock this? https://github.com/lightningdevkit/rust-lightning/pull/3412

**Reply by elsirion** [reply]
Sound useful high-level, thx for making me aware!
